## Data cleaning

In [341]:
import pandas as pd

%store -r data_frame
%store -r list_of_columns
%store -r missing_values_above_80_percent
%store -r missing_values_below_10_percent 

#remove limit displaying rows and columns
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

### Delete columns and rows with missing values.

Removal of columns with more than 80% missing values.

In [342]:
data_frame = data_frame.drop(columns=missing_values_above_80_percent)
data_frame.head()

id                         listing_url       scrape_id last_scraped  \
0   2384   https://www.airbnb.com/rooms/2384  20220914061546   2022-09-14   
1   7126   https://www.airbnb.com/rooms/7126  20220914061546   2022-09-14   
2  10945  https://www.airbnb.com/rooms/10945  20220914061546   2022-09-14   
3  12140  https://www.airbnb.com/rooms/12140  20220914061546   2022-09-14   
4  24833  https://www.airbnb.com/rooms/24833  20220914061546   2022-09-14   

        source                                               name  \
0  city scrape                       Hyde Park - Walk to UChicago   
1  city scrape                Tiny Studio Apartment 94 Walk Score   
2  city scrape                              The Biddle House (#1)   
3  city scrape                           Lincoln Park Guest House   
4  city scrape  Prime LincolnPark 1 Block Fullerton Express L ...   

                                         description  \
0  You are invited to be the sole Airbnb guest in...   
1  A very small studio in a wonderful neighborhoo...   
2  Beautiful first floor apartment in Historic Ol...   
3  Bed and Breakfast license issued by the City o...   
4  Perfect private entrance apartment one block f...   

                               neighborhood_overview  \
0  The apartment is less than one block from beau...   
1  Ukrainian Village was just named "Hottest Neig...   
2                                                NaN   
3  The Guest House is in the heart of Chicago's m...   
4  Lincoln Park is a great neighborhood where peo...   

                                         picture_url  host_id  \
0  https://a0.muscache.com/pictures/acf6b3c0-47f2...     2613   
1  https://a0.muscache.com/pictures/51073/16c81c7...    17928   
2  https://a0.muscache.com/pictures/58d1a420-a24b...    33004   
3  https://a0.muscache.com/pictures/55126/6a4f01b...    46734   
4  https://a0.muscache.com/pictures/ac17be6f-40bd...   101521   

                                   host_url          host_name  host_since  \
0    https://www.airbnb.com/users/show/2613            Rebecca  2008-08-29   
1   https://www.airbnb.com/users/show/17928              Sarah  2009-05-19   
2   https://www.airbnb.com/users/show/33004        At Home Inn  2009-08-21   
3   https://www.airbnb.com/users/show/46734  Sharon And Robert  2009-10-18   
4  https://www.airbnb.com/users/show/101521                Red  2010-03-31   

  host_location                                         host_about  \
0   Chicago, IL  My 2 bdrm apartment is a 2nd floor walk-up in ...   
1   Chicago, IL  We live in Chicago. We love to travel and see ...   
2   Chicago, IL  Hi, we're Bob and Liz Biddle, long time Chicag...   
3   Chicago, IL  Best in the city with long-time urban-dwelling...   
4   Chicago, IL  Travel is fun and exciting but can be hard on ...   

   host_response_time host_response_rate host_acceptance_rate  \
0      within an hour               100%                  97%   
1      within an hour               100%                  96%   
2      within an hour               100%                  99%   
3      within an hour               100%                  89%   
4  within a few hours                90%                  77%   

  host_is_superhost                                 host_thumbnail_url  \
0                 t  https://a0.muscache.com/im/pictures/user/8bb47...   
1                 t  https://a0.muscache.com/im/users/17928/profile...   
2                 t  https://a0.muscache.com/im/pictures/user/e23d4...   
3                 f  https://a0.muscache.com/im/users/46734/profile...   
4                 f  https://a0.muscache.com/im/users/101521/profil...   

                                    host_picture_url host_neighbourhood  \
0  https://a0.muscache.com/im/pictures/user/8bb47...          Hyde Park   
1  https://a0.muscache.com/im/users/17928/profile...  Ukrainian Village   
2  https://a0.muscache.com/im/pictures/user/e23d4...           Old Town   
3  https://a0.muscache.com/im/users/4

Removal of blank rows in columns where the missing value was less than 10%

In [343]:
data_frame = data_frame.dropna(subset=missing_values_below_10_percent)

In [344]:
data_frame.shape

(5722, 72)

After removing the missing data, 5722 rows and 72 columns remain.

In [345]:
data_frame.describe()

id     scrape_id       host_id  host_listings_count  \
count  5.722000e+03  5.722000e+03  5.722000e+03          5722.000000   
mean   1.823850e+17  2.022091e+13  1.507197e+08           300.825061   
std    2.888265e+17  0.000000e+00  1.437020e+08          1033.410992   
min    2.384000e+03  2.022091e+13  2.153000e+03             1.000000   
25%    2.987785e+07  2.022091e+13  2.800625e+07             1.000000   
50%    4.786159e+07  2.022091e+13  1.074344e+08             4.000000   
75%    5.718707e+17  2.022091e+13  2.430356e+08            18.000000   
max    7.149113e+17  2.022091e+13  4.778108e+08          4050.000000   

       host_total_listings_count     latitude    longitude  accommodates  \
count                5722.000000  5722.000000  5722.000000   5722.000000   
mean                  338.379588    41.893914   -87.663783      4.519923   
std                  1139.895562     0.062039     0.044977      3.071889   
min                     1.000000    41.650640   -87.901560      1.000000   
25%                     2.000000    41.863507   -87.687792      2.000000   
50%                     6.000000    41.897823   -87.659170      4.000000   
75%                    27.000000    41.938038   -87.631996      6.000000   
max                  6605.000000    42.033620   -87.534090     16.000000   

          bedrooms         beds  minimum_nights  maximum_nights  \
count  5722.000000  5722.000000     5722.000000     5722.000000   
mean      1.875393     2.349528       13.328207      648.540196   
std       1.139014     1.784929       43.400135      475.750359   
min       1.000000     1.000000        1.000000        1.000000   
25%       1.000000     1.000000        2.000000      149.250000   
50%       2.000000     2.000000        2.000000      365.000000   
75%       2.000000     3.000000       32.000000     1125.000000   
max      12.000000    20.000000     1125.000000     1125.000000   

       minimum_minimum_nights  maximum_minimum_nights  minimum_maximum_nights  \
count             5722.000000             5722.000000             5722.000000   
mean                14.832226               63.054002              822.465921   
std                 54.337958              188.154185              447.104026   
min                  1.000000                1.000000                1.000000   
25%                  2.000000                2.000000              365.000000   
50%                  2.000000                3.000000             1125.000000   
75%                 32.000000               32.000000             1125.000000   
max               1125.000000             1125.000000             1125.000000   

       maximum_maximum_nights  minimum_nights_avg_ntm  maximum_nights_avg_ntm  \
count            5.722000e+03             5722.000000            5.722000e+03   
mean             3.378606e+06               60.974956            2.636483e+06   
std              8.510854e+07              182.296170            6.824100e+07   
min              2.000000e+00                1.000000            2.000000e+00   
25%              3.650000e+02                2.000000            3.650000e+02   
50%              1.125000e+03                2.800000            1.125000e+03   
75%              1.125000e+03               32.000000            1.125000e+03   
max              2.147484e+09             1125.000000            2.118332e+09   

       availability_30  availability_60  availability_90  availability_365  \
count      5722.000000      5722.000000      5722.000000       5722.000000   
mean         11.461028        30.198357        51.967669        212.767913   
std           9.333101        19.331332        28.867650        125.569926   
min           0.000000         0.000000         0.000000          0.000000   
25%           3.000000        13.000000        32.000000         86.000000   
50%          10.000000        33.000000        60.000000        243.000000   
75%          19.000000        47.000000        75.750000        335.00000

In [346]:
data_frame.head()

id                         listing_url       scrape_id last_scraped  \
0   2384   https://www.airbnb.com/rooms/2384  20220914061546   2022-09-14   
1   7126   https://www.airbnb.com/rooms/7126  20220914061546   2022-09-14   
2  10945  https://www.airbnb.com/rooms/10945  20220914061546   2022-09-14   
3  12140  https://www.airbnb.com/rooms/12140  20220914061546   2022-09-14   
4  24833  https://www.airbnb.com/rooms/24833  20220914061546   2022-09-14   

        source                                               name  \
0  city scrape                       Hyde Park - Walk to UChicago   
1  city scrape                Tiny Studio Apartment 94 Walk Score   
2  city scrape                              The Biddle House (#1)   
3  city scrape                           Lincoln Park Guest House   
4  city scrape  Prime LincolnPark 1 Block Fullerton Express L ...   

                                         description  \
0  You are invited to be the sole Airbnb guest in...   
1  A very small studio in a wonderful neighborhoo...   
2  Beautiful first floor apartment in Historic Ol...   
3  Bed and Breakfast license issued by the City o...   
4  Perfect private entrance apartment one block f...   

                               neighborhood_overview  \
0  The apartment is less than one block from beau...   
1  Ukrainian Village was just named "Hottest Neig...   
2                                                NaN   
3  The Guest House is in the heart of Chicago's m...   
4  Lincoln Park is a great neighborhood where peo...   

                                         picture_url  host_id  \
0  https://a0.muscache.com/pictures/acf6b3c0-47f2...     2613   
1  https://a0.muscache.com/pictures/51073/16c81c7...    17928   
2  https://a0.muscache.com/pictures/58d1a420-a24b...    33004   
3  https://a0.muscache.com/pictures/55126/6a4f01b...    46734   
4  https://a0.muscache.com/pictures/ac17be6f-40bd...   101521   

                                   host_url          host_name  host_since  \
0    https://www.airbnb.com/users/show/2613            Rebecca  2008-08-29   
1   https://www.airbnb.com/users/show/17928              Sarah  2009-05-19   
2   https://www.airbnb.com/users/show/33004        At Home Inn  2009-08-21   
3   https://www.airbnb.com/users/show/46734  Sharon And Robert  2009-10-18   
4  https://www.airbnb.com/users/show/101521                Red  2010-03-31   

  host_location                                         host_about  \
0   Chicago, IL  My 2 bdrm apartment is a 2nd floor walk-up in ...   
1   Chicago, IL  We live in Chicago. We love to travel and see ...   
2   Chicago, IL  Hi, we're Bob and Liz Biddle, long time Chicag...   
3   Chicago, IL  Best in the city with long-time urban-dwelling...   
4   Chicago, IL  Travel is fun and exciting but can be hard on ...   

   host_response_time host_response_rate host_acceptance_rate  \
0      within an hour               100%                  97%   
1      within an hour               100%                  96%   
2      within an hour               100%                  99%   
3      within an hour               100%                  89%   
4  within a few hours                90%                  77%   

  host_is_superhost                                 host_thumbnail_url  \
0                 t  https://a0.muscache.com/im/pictures/user/8bb47...   
1                 t  https://a0.muscache.com/im/users/17928/profile...   
2                 t  https://a0.muscache.com/im/pictures/user/e23d4...   
3                 f  https://a0.muscache.com/im/users/46734/profile...   
4                 f  https://a0.muscache.com/im/users/101521/profil...   

                                    host_picture_url host_neighbourhood  \
0  https://a0.muscache.com/im/pictures/user/8bb47...          Hyde Park   
1  https://a0.muscache.com/im/users/17928/profile...  Ukrainian Village   
2  https://a0.muscache.com/im/pictures/user/e23d4...           Old Town   
3  https://a0.muscache.com/im/users/4

Updating lists of columns of a given type after deleting NaN values

In [347]:
int_columns = [*data_frame.select_dtypes(include=['int'])]
float_columns = [*data_frame.select_dtypes(include=['float'])]
object_columns = [*data_frame.select_dtypes(include=['object'])]

### Verify that all columns have the correct types

In [348]:
data_frame[['host_response_rate','price','host_acceptance_rate']].head()


host_response_rate    price host_acceptance_rate
0               100%  $110.00                  97%
1               100%   $91.00                  96%
2               100%  $157.00                  99%
3               100%  $329.00                  89%
4                90%   $86.00                  77%

In [349]:
'host_response_rate' in object_columns, 'price' in object_columns, 'host_acceptance_rate' in object_columns

(True, True, True)

The above columns should be cleared of "%" and $ signs, because they were incorrectly qualified as 'object'

In [350]:
data_frame.price = data_frame.price.apply(lambda x: x.replace('$',''))
data_frame.price = data_frame.price.apply(lambda x: x.replace(',',''))
data_frame.price = pd.to_numeric(data_frame.price, errors='coerce')
data_frame.price.describe()

count     5722.000000
mean       219.076896
std        278.169271
min         18.000000
25%         96.000000
50%        156.000000
75%        249.000000
max      10000.000000
Name: price, dtype: float64

The 'price' column has been cleaned up and its type changed to float64

In [351]:
data_frame.host_response_rate = data_frame.host_response_rate.str.replace('%','')
data_frame.host_response_rate = data_frame.host_response_rate.astype(float)/100

data_frame.host_acceptance_rate = data_frame.host_acceptance_rate.str.replace('%','')
data_frame.host_acceptance_rate = data_frame.host_acceptance_rate.astype(float)/100

In [352]:
data_frame[['host_response_rate','price','host_acceptance_rate']].dtypes

host_response_rate      float64
price                   float64
host_acceptance_rate    float64
dtype: object

Removed the '%' signs and changed the other two columns to float, reflecting percentages to decimals

### Check where there are still missing values:

In [353]:
missing_values = {}

def missing_values_check(columns_to_check):
    #check if missing values dcit is empty
    missing_values.clear()
    
    for index in columns_to_check:
        if data_frame[index].isna().sum() > 0:
            missing_values[index] = data_frame[index].isna().sum()
    
    if len(missing_values) == 0:
        print ('There are no missing values ​​in the following columns:' + f"{columns_to_check}")
    else:
        print(missing_values)

missing_values_check(object_columns)

{'neighborhood_overview': 1723, 'host_location': 842, 'host_about': 1769, 'host_response_time': 191, 'host_response_rate': 191, 'neighbourhood': 1723, 'first_review': 934, 'last_review': 934, 'license': 756}


In [354]:
missing_values_col_indexes = list(missing_values.keys())
data_frame[missing_values_col_indexes].dtypes

neighborhood_overview     object
host_location             object
host_about                object
host_response_time        object
host_response_rate       float64
neighbourhood             object
first_review              object
last_review               object
license                   object
dtype: object

Missing values ​​should be supplemented, e.g. with average values ​​or the most numerous categorical ones

In [355]:
data_frame[missing_values_col_indexes].head()

neighborhood_overview host_location  \
0  The apartment is less than one block from beau...   Chicago, IL   
1  Ukrainian Village was just named "Hottest Neig...   Chicago, IL   
2                                                NaN   Chicago, IL   
3  The Guest House is in the heart of Chicago's m...   Chicago, IL   
4  Lincoln Park is a great neighborhood where peo...   Chicago, IL   

                                          host_about  host_response_time  \
0  My 2 bdrm apartment is a 2nd floor walk-up in ...      within an hour   
1  We live in Chicago. We love to travel and see ...      within an hour   
2  Hi, we're Bob and Liz Biddle, long time Chicag...      within an hour   
3  Best in the city with long-time urban-dwelling...      within an hour   
4  Travel is fun and exciting but can be hard on ...  within a few hours   

   host_response_rate                     neighbourhood first_review  \
0                 1.0  Chicago, Illinois, United States   2015-01-09   
1                 1.0  Chicago, Illinois, United States   2009-07-03   
2                 1.0                               NaN   2014-04-28   
3                 1.0  Chicago, Illinois, United States   2015-06-29   
4                 0.9  Chicago, Illinois, United States   2010-05-24   

  last_review                    license  
0  2022-09-09               R17000015609  
1  2022-08-28               R21000075737  
2  2022-08-23                    2209984  
3  2022-09-12               R20000055258  
4  2022-04-23  City registration pending

Most of the missing values ​​are contained in the 'object' type. Columns: host_location, host_response_time, neighborhood, will be filled with the most numerous categories. The host_response_rate column will be filled with an average value. In the remaining columns we will replace NaN with the string 'no info'. From the point of view of further analysis, the missing description 'host_about' will probably not be a problem, but it will be disadvantageous for us to remove 1769 records including NaN for this reason.

In [356]:
to_fill_list = ['host_location', 'host_response_time', 'neighbourhood']

A function that fills the columns in the list above with the most common values:

In [357]:
def fill_missing(col_name):
    result = data_frame.groupby(col_name)[col_name].count().rename('count')
    result_df = pd.DataFrame(result)
    result_df = result_df.reset_index().sort_values(by=['count'],ascending=False)
    to_fill = result_df[col_name].iloc[0]
    data_frame[col_name] = data_frame[col_name].fillna(value=to_fill)
    
for col in to_fill_list:
    fill_missing(col)



Checking if columns containing NaN values ​​have been completed:

In [361]:

missing_values_check(to_fill_list)

There are no missing values ​​in the following columns:['host_location', 'host_response_time', 'neighbourhood']


In [365]:
data_frame.host_acceptance_rate.mean()

0.9124379587556799

Poszukujemy odstających wartości

In [359]:
data_frame.accommodates.max(), data_frame.accommodates.mean()

(16, 4.519923103809857)

Usuwamy odstające wartości

Sprawdzamy gdzie jeszcze mamy nany i uzupełniamy wartościami średnimi lub najliczniejszymi kategoriami

Jak nie ma nanów tworzymy rozkład normalny dla int i float, pewnie jeszcze raz bierzemy float